### ISSUE TITLE
line() produces artifacts with shift >= 2 and thickness >= 2 for almost zero length lines

### ISSUE THREAD

**Issue by ssmir** [opencv/opencv#7927]
##### System information (version)

- OpenCV => 3.2.0-dev, revision 7dd3723
- Operating System / Platform => macOS Sierra 10.12.2
- Compiler => Apple LLVM version 8.0.0 (clang-800.0.42.1)

##### Detailed description

When I use cv::line() with shift parameter >= 2 and thickness >= 2 to draw lines of almost zero length, I get artifacts on an image. In C++ it is a pixel with the same Y coordinate as input and zero X. In Python it is a horisontal line with the same Y as input and with somewhat random length usually equal to full image width.

I did not have this problem with OpenCV 2.4.13.2 or with 3.0.0 release.

##### Steps to reproduce
Expected to draw a small cross in the center. But it also draws a white pixel with coordinates (0, 100)
```.cpp
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

using namespace cv;

int main( void ){
    Mat img = Mat::zeros(200, 200, CV_8U);

    // shift >= 2 and thickness >= 2 are essential
    line(img, Point(400, 400), Point(400, 401), 255, 1, LINE_8, 2);

    imshow( "w1", img );
    moveWindow( "w1", 0, 200 );
    waitKey( 0 );
    return(0);
}
```

Expected to draw a small cross in the center. But also produces a horisontal line with Y=100 in the center of the image.
```.py
import cv2
import numpy as np

img = np.zeros((200, 200), dtype=np.uint8)

cv2.line(img, (400, 400), (400, 401), 255, thickness=2, shift=2)

cv2.imshow('image', img)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

**Comment by paroj** [comment]
also noticed that. I think it is related to this PR: https://github.com/opencv/opencv/pull/7812

**Comment by paroj** [comment]
probably fixed by #7959

**Comment by ssmir** [comment]
Yes, no such problem in 45f0cf0. Thanks.
